### PR TITLE
Added absolute order to EntityType Keys, Foreign Keys, Indexes, and Navigations

### PR DESCRIPTION
**Pull request by mikary**
Adding total order to address #1314 and fixing tests broken by this change.
@AndriySvyryd @bricelam @lajones

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/EntityType.cs [hunk 1]
```diff
@@ -197,7 +197,7 @@ public virtual Key AddKey([NotNull] IReadOnlyList<Property> properties)
         {
             Check.NotEmpty(properties, "properties");
 
-            var key = TryGetKey(properties);
+            var key = TryGetPrimaryKey(properties);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: Was this an intentional change?

Comment (mikary) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: yes, TryGetKey searches first for a primary key and then through the ```_keys``` This preserves the logic but still lets us use the BinarySearch

Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: I see. Why not change the TryGet methods to use binary search instead?

Comment (mikary) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: The BinarySearch requires an instance, we use it in some of the Properties methods, but we have a performance note about the impact of creating a surrogate instance for the sake of searching.

We have some options to consider with respect to surrogates, binary search, and lazy if we want to start down that path on this PR.

Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: Yes, we should measure the impact of this at some point, but for now we should be consistent. Either use binary search in all TryGet methods or in none.

File: test/EntityFramework.Core.Tests/Metadata/EntityTypeTest.cs [hunk 2]
```diff
@@ -179,7 +179,7 @@ public void Can_add_and_get_a_key()
 
             Assert.NotNull(key2);
             Assert.Same(key2, entityType.GetKey(idProperty));
-            Assert.Equal(new[] { key1, key2 }, entityType.Keys.ToArray());
+            Assert.Equal(new[] { key2, key1 }, entityType.Keys.ToArray());
```
Comment (lajones) [inline test/EntityFramework.Core.Tests/Metadata/EntityTypeTest.cs]: Nit: this test looks like it belongs in Keys_are_ordered_by_property_count_then_property_names() below. Here would just add some non-overlapping keys and ensure you can retrieve them.

File: test/EntityFramework.Core.Tests/Metadata/EntityTypeTest.cs [hunk 3]
```diff
@@ -269,6 +269,20 @@ public void Removing_a_key_throws_if_it_referenced_from_a_foreign_key_in_the_mod
...
+        {
+            var customerType = new EntityType(typeof(Customer), new Model());
+            var idProperty = customerType.GetOrAddProperty(Customer.IdProperty);
+            var nameProperty = customerType.GetOrAddProperty(Customer.NameProperty);
+
+            var k2 = customerType.GetOrAddKey(nameProperty);
+            var k3 = customerType.GetOrAddKey(new[] { idProperty, nameProperty });
+            var k1 = customerType.GetOrAddKey(idProperty);
+
+            Assert.True(new[] { k1, k2, k3 }.SequenceEqual(customerType.Keys));
```
Comment (lajones) [inline test/EntityFramework.Core.Tests/Metadata/EntityTypeTest.cs]: Nit: maybe add another key k4 on (idProperty, abcProperty) just to prove that even in composite keys it orders them by name.

**Comment by lajones** [comment]
:shipit:

**Comment by AndriySvyryd** [comment]
:shipit:

**Comment by mikary** [comment]
Updated for PR feedback, there are enough changes that it could use another look before I merge.

**Comment by lajones** [comment]
:shipit:

**Comment by AndriySvyryd** [comment]
nice :shipit:

**Pull request merged**